### ISSUE TITLE
RuntimeError: iter.device(arg).is_cuda() INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1607370116979/work/aten/src/ATen/native/cuda/Loops.cuh":94, please report a bug to PyTorch. 

### ISSUE THREAD

**Issue by yongshuo-Z** [pytorch/pytorch#54487]
## 🐛 Bug

Hi, when I am running the code,  I met this bug: RuntimeError: iter.device(arg).is_cuda() INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1607370116979/work/aten/src/ATen/native/cuda/Loops.cuh":94, please report a bug to PyTorch. 

## To Reproduce

1. my code is available [here](https://drive.google.com/file/d/19_Jeqa889ue3JRNYMIVyfrE3eHVWgwJu/view?usp=sharing). Just simply run `main.ipynb`.
2. I'm trying to reproduce a specific triplet learning algorithm from this paper: https://arxiv.org/abs/2007.11186. For training, the code runs well on the CPU, but when switched to GPU (both single/data-parallel don't work), this error occurs.

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-cf6df666728d> in <module>()
      6         output1, output2, output3 = model(anchor, positive, negative)
      7         optimizer.zero_grad()
----> 8         loss = loss_fn(output1, output2, output3)
      9         loss.backward()
     10         optimizer.step()

/opt/anaconda3/envs/cryo-torch/lib/python3.6/site-packages/torch/nn/modules/module.py in _call_impl(self, *input, **kwargs)
    725             result = self._slow_forward(*input, **kwargs)
    726         else:
--> 727             result = self.forward(*input, **kwargs)
    728         for hook in itertools.chain(
    729                 _global_forward_hooks.values(),

/root/cryo/code/self-supervised/triplet/losses.py in forward(self, anchor, positive, negative, size_average)
     18         Dap = F.pairwise_distance(anchor, positive, p=2).pow(2)
     19         Dan = F.pairwise_distance(anchor, negative, p=2).pow(2)
---> 20         Lst = torch.max(torch.tensor(0), Dap - Dan + self.margin1).squeeze()
     21 
     22         new_neg = negative.reshape([1, 128, 1, 1])

RuntimeError: iter.device(arg).is_cuda() INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1607370116979/work/aten/src/ATen/native/cuda/Loops.cuh":94, please report a bug to PyTorch. 

## Expected behavior

I wish to train the network on GPU as it works on CPU. It seems something wrong with the Loss function part I wrote? I am new to it, any help will be greatly appreciated.

## Environment

 - PyTorch Version 1.7.1:
 - OS Ubuntu:
 - How you installed PyTorch  conda:
 - Python 3.6:
 - CUDA/cuDNN 10.1:
 - GPU GeForce 2080, 11GB

Thanks in advance!

**Comment by skyline75489** [comment]
Please fill the context needed. There's [collect_env.py](https://github.com/pytorch/pytorch/blob/master/torch/utils/collect_env.py) that helps you do that.

**Comment by yongshuo-Z** [comment]
> Please fill the context needed. There's [collect_env.py](https://github.com/pytorch/pytorch/blob/master/torch/utils/collect_env.py) that helps you do that.

Sure. updated.

**Comment by bdhirsh** [comment]
@yongshuo-Z Can you try to post a a minimum-repro of the error? The error looks like it's happening in your loss function: `Lst = torch.max(torch.tensor(0), Dap - Dan + self.margin1).squeeze()`. But it's hard to tell what the values of the inputs are that are triggering the internal assert. It would be helpful if you could determine what the values of `Dap/Dan/self.margin1` are when the internal assert is triggered on that line.

At a quick glance though, try changing `torch.tensor(0)` on that line to `torch.tensor(0, device=device)`. The kernel probably isn't happy because that tensor will by default go to the cpu, but your other inputs are on cuda.

It would still be nice to get a minimum repro to fix the internal assert though!

**Comment by yongshuo-Z** [comment]
> @yongshuo-Z Can you try to post a a minimum-repro of the error? The error looks like it's happening in your loss function: `Lst = torch.max(torch.tensor(0), Dap - Dan + self.margin1).squeeze()`. But it's hard to tell what the values of the inputs are that are triggering the internal assert. It would be helpful if you could determine what the values of `Dap/Dan/self.margin1` are when the internal assert is triggered on that line.
> 
> At a quick glance though, try changing `torch.tensor(0)` on that line to `torch.tensor(0, device=device)`. The kernel probably isn't happy because that tensor will by default go to the cpu, but your other inputs are on cuda.
> 
> It would still be nice to get a minimum repro to fix the internal assert though!

Thanks!! I changed `torch.tensor(0)` to `torch.tensor(0, device=device)` and change `fcl = nn.Conv2d(128, 1, 1, bias = False)` to `fcl = nn.Conv2d(128, 1, 1, bias = False).to(self.device)`. Then it works!

If you still need a minimum repro, I will post it later.

**Comment by yongshuo-Z** [comment]
> @yongshuo-Z Can you try to post a a minimum-repro of the error? The error looks like it's happening in your loss function: `Lst = torch.max(torch.tensor(0), Dap - Dan + self.margin1).squeeze()`. But it's hard to tell what the values of the inputs are that are triggering the internal assert. It would be helpful if you could determine what the values of `Dap/Dan/self.margin1` are when the internal assert is triggered on that line.
> 
> At a quick glance though, try changing `torch.tensor(0)` on that line to `torch.tensor(0, device=device)`. The kernel probably isn't happy because that tensor will by default go to the cpu, but your other inputs are on cuda.
> 
> It would still be nice to get a minimum repro to fix the internal assert though!

Thanks!! I changed `torch.tensor(0)` to `torch.tensor(0, device=device)` and change `fcl = nn.Conv2d(128, 1, 1, bias = False)` to `fcl = nn.Conv2d(128, 1, 1, bias = False).to(self.device)`. Then it works!

If you still need a minimum repro, I will post it later.

**Comment by ngimel** [comment]
Yes please, minimum repro is still helpful so that we can improve error message.
